File: 460_lfu_cache.py
class LFUCache:
    """
        Time Complexity - O(1)
    """

    class Node:

        def __init__(self, key: int, val: int):
            self.key = key
            self.val = val
            self.count = 1
            self.next = None
            self.prev = None

    class DLList:

        def __init__(self):
            self.head = LFUCache.Node(None, None)
            self.tail = LFUCache.Node(None, None)
            self.head.next = self.tail
            self.tail.prev = self.head
            self.size = 0

        def add_to_head(self, node) -> None:
            node.next = self.head.next
            node.prev = self.head
            self.head.next.prev = node
            self.head.next = node
            self.size += 1

        def remove_node(self, node) -> None:
            node.prev.next = node.next
            node.next.prev = node.prev
            self.size -= 1

        def remove_last(self):
            last_node = self.tail.prev
            self.remove_node(last_node)
            return last_node

    def __init__(self, capacity: int):
        self.dic = {}
        self.freq_dic = {}
        self.cache_size = 0
        self.capacity = capacity
        self.min = float('-inf')

    def get(self, key: int) -> int:
        # exists
        if key in self.dic:
            node = self.dic[key]
            self.update(node)
            return node.val
        # does not exists
        return -1

    def put(self, key: int, value: int) -> None:
        # edge case
        if self.capacity == 0:
            return

        # Case 1: Key is already there in cache
        if key in self.dic:
            node = self.dic[key]
            node.val = value
            self.update(node)
        else:
            # Case 2: Key not present
            node = self.Node(key, value)
            # Cache is full
            if self.cache_size == self.capacity:
                # get the DLL with minimum value
                min_list = self.freq_dic[self.min]
                to_be_removed = min_list.remove_last()
                self.dic.pop(to_be_removed.key)
                self.cache_size -= 1
            if 1 not in self.freq_dic:
                self.freq_dic[1] = self.DLList()
            self.freq_dic[1].add_to_head(node)
            self.dic[key] = node
            self.min = 1
            self.cache_size += 1

    # updating node in freq map
    def update(self, node) -> None:
        # get the DLL with this count
        old_list = self.freq_dic[node.count]
        old_list.remove_node(node)
        if node.count == self.min and old_list.size == 0:
            self.min += 1
        node.count += 1
        if node.count not in self.freq_dic:
            self.freq_dic[node.count] = self.DLList()
        self.freq_dic[node.count].add_to_head(node)
# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 460_LFU_Cache.java
    /*  Explanation
    # Leetcode problem link : https://leetcode.com/problems/lfu-cache/
    Time Complexity for operators : o(1) 
    Extra Space Complexity for operators : o(1)
    Did this code successfully run on Leetcode : NA
    Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
        # Basic approach : 
        # Optimized approach: 
                              
            # 1. 
                    A) Initialize the variables keyToCount, countToLRU, capacity and min_count.
                    B) Create function getLHM(), it will create new LinkedHashMap.
                    C) get() - Check if key present in keyToCount if no then return -1.
                             - Then if present then get the count from keyToCount and then find that in countToLRU's Linkedhashmap
                               to get the value. Remove this from LinkedHashMap as counter is now changing for this.
                             - If this count == min_count and countToLRU is emoty then increse the min_count.
                             - Update the keyToCount with new counter value.
                             - Update countToLRU with new counter value.
                    D) put() - If capacity <= 0 then return.
                             - If element is present then do the get(), (C) step on it so that it will get incresed its coounts 
                               and then added into the respective haqshmap.
                             - If capacity is reached then, get the key of first ekement in the linkedhashmap of min_count.
                               remove it from LinkedHashmap and alsoo from KeyToCount.
                             - If it passes all these checks, then put the element regulary by updating both the hashmap with count.
    */

class LFUCache {

    HashMap<Integer, Integer> keyToCount = new HashMap<>(); // (key, count)
    HashMap<Integer, LinkedHashMap<Integer, Integer>> countToLRU = new HashMap<>();//(count, (key, value))
    
    int min_count = 0;
    int capacity = 0;
    
    private LinkedHashMap<Integer, Integer> getLHM(){
        return new LinkedHashMap<Integer, Integer>(capacity);
    }
    
    public LFUCache(int capacity) {
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if(!keyToCount.containsKey(key))
            return -1;
        
        int count = keyToCount.get(key);
        int value = countToLRU.get(count).remove(key); // it returns the value
        
        if(count == min_count && countToLRU.get(count).isEmpty()){
            min_count += 1;
        }
        
        keyToCount.put(key, count+1);
        countToLRU.putIfAbsent(count + 1, getLHM());
        
        countToLRU.get(count+1).put(key, value);
        
        return value;
    }
    
    public void put(int key, int value) {
        if(this.capacity<=0)
            return;
        
        // put if element is present
        if(keyToCount.containsKey(key)){
            get(key);
            
            int count = keyToCount.get(key);
            countToLRU.get(count).put(key, value);
            
            return;
        }
        
        // capacity is reached 
        if(this.capacity <= keyToCount.size()){
            int temp_key = countToLRU.get(min_count).entrySet().iterator().next().getKey();
            countToLRU.get(min_count).remove(temp_key);
            keyToCount.remove(temp_key);
        }
        
        // put regulary if element is not present
        min_count = 1;
        keyToCount.put(key,1);
        countToLRU.putIfAbsent(1, getLHM());   
        countToLRU.get(1).put(key, value);
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */